### PR TITLE
First partial files set.

### LINKED ISSUE #3: Migrate WPN Code
Copy over files from 4.0 and update them to become an extension

### PR DESCRIPTION
**Pull request by rxu**
Creating this as a start point place to collaborate on.

Closes #3

### CODE REVIEW

File: styles/all/template/webpush.js [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+'use strict';
+
+/**
+ * Convert raw key ArrayBuffer to base64 string.
+ *
+ * @param {ArrayBuffer} rawKey Raw key array buffer as exported by SubtleCrypto exportKey()
+ * @returns {string} Base64 encoded raw key string
+ */
+phpbb.rawKeyToBase64 = (rawKey) => {
```
Comment (iMattPro) [inline styles/all/template/webpush.js:11]: Do these base64 functions need to be in the ACP JS file instead? It looks like they're only used in that file, and I'm not sure they will be accessible in the ACP if they are in this file.

**Comment by iMattPro** [comment]
I think the ACP can be done just using the event listener to inject its settings into acp_board (just saying it shouldn't need to be a whole stand alone ACP module). It's only 3 settings I think. Looks like Marc added WebPush to the Client Communication section of ACP board settings.

File: config/tables.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+phpbb.webpushnotifications.tables.notification_push: '%core.table_prefix%webpushnotifications_notification_push'
```
Comment (iMattPro) [inline config/tables.yml:1]: would this be better?
`phpbb.wpn.tables.notification_push: '%core.table_prefix%wpn_notification_push`

**Comment by rxu** [comment]
> I think the ACP can be done just using the event listener to inject its settings into acp_board (just saying it shouldn't need to be a whole stand alone ACP module). It's only 3 settings I think. Looks like Marc added WebPush to the Client Communication section of ACP board settings.

The problem here is that you can only add new configuration options to already existing ACP modes, but there's no event to add new custom modes so you get this
```php
			default:
				trigger_error('NO_MODE', E_USER_ERROR);
			break;
```

Shall we add new event and restrict this extension to phpBB >= 3.3.12 or create its own ACP module anyway?

**Comment by iMattPro** [comment]
Let’s just make a simple ACP MODULE where all the logic is inside the module php file (let’s not overcompensate it with controllers). I guess that would have less chance of conflict too inside of phpBB 4

**Comment by iMattPro** [comment]
But just thinking (and not trying it) can't we create a module like Marc did, via the _info file and migration, and then once that module exists in the sidebar, wouldn't the mode work? Couldn't we then do an event like:

File: adm/style/event/acp_overall_footer_body_after.html [hunk 3]
```diff
@@ -0,0 +1 @@
+{% INCLUDEJS '@phpbb_webpushnotifications/webpushnotifications_acp.js.js' %}
```
Comment (iMattPro) [inline adm/style/event/acp_overall_footer_body_after.html:1]: too many `.js` at the end ;)

File: config/tables.yml (hunk 2 above)
Comment (rxu) [reply config/tables.yml:1]: Or just `webpush`? What is more beautiful? :)

File: acp/wpn_acp_module.php [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ * phpBB Browser Push Notifications. An extension for the phpBB Forum Software package.
+ *
+ * @copyright (c) 2023, phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ */
+
+namespace phpbb\webpushnotifications\acp;
+
+use Minishlink\WebPush\VAPID;
```
Comment (iMattPro) [inline acp/wpn_acp_module.php:13]: do we need this line?

File: language/en/info_acp_webpushnotifications.php [hunk 5]
```diff
@@ -38,8 +38,17 @@
 //
 
 $lang = array_merge($lang, [
-	'ACP_WEBPUSH_SETTINGS'		=> 'Web Push settings',
-	'LOG_CONFIG_WEBPUSH'		=> '<strong>Altered Web Push settings</strong>',
+	'ACP_WEBPUSH_SETTINGS'			=> 'Web Push settings',
```
Comment (iMattPro) [inline language/en/info_acp_webpushnotifications.php:41]: Only thing about this, is info_acp language files are automatically loaded on every single page in the ACP, so this file should only contain lang vars we need accessible everywhere in the ACP, like LOG vars and other global lang vars.

The Lang vars that will only be used in our module should normally be kept inside their own lang file so they are only used when needed.

File: styles/all/template/wpn_acp.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/* global phpbb */
```
Comment (iMattPro) [inline styles/all/template/wpn_acp.js:1]: This JS file, if only used in the ACP, could be put in the arm/style folder instead

File: adm/style/event/acp_overall_footer_body_after.html [hunk 7]
```diff
@@ -0,0 +1 @@
+{% INCLUDEJS '@phpbb_webpushnotifications/wpn_acp.js' %}
```
Comment (iMattPro) [inline adm/style/event/acp_overall_footer_body_after.html:1]: this line should be in our adm/style/wpn_acp_settings.html page now instead

**Comment by rxu** [comment]
It seems that this ext will be restricted to phpBB 3.3.12 as it seems to be requiring new template event in `ucp_notifications.html` to implement webpush `Subscribe` / `Unsubscribe` buttons.

**Comment by iMattPro** [comment]
> It seems that this ext will be restricted to phpBB 3.3.12 as it seems to be requiring new template event in `ucp_notifications.html` to implement webpush `Subscribe` / `Unsubscribe` buttons. So we need to create that event.

That's fine. Just means we will have plenty of time to finish this extension then ;)

**Comment by iMattPro** [comment]
One thing I noticed with the ACP. You can't save settings unless you always have those API key fields filled in correctly, otherwise you always get an error and all the fields reset. So, if they are blank, you can't just simply set enable/disable setting. Maybe it should only save settings for Enabling Push feature if those inputs have valid data, but when disabling the feature, allow those input fields to be empty too?

**Comment by rxu** [comment]
`Subscribe` button is always disabled and `Unsubscribe` one is not rendered for me currently on localhost. Looks like due to `webpush.js` secure context detection issues but the same thing on Marc's branch works fine.

**Comment by rxu** [comment]
Ah, disregard. Now works after your `[Fix some broken links](https://github.com/phpbb-extensions/webpushnotifications/pull/2/commits/7e13a32600fda37403134e6fcc392cf903e35b35)` commit.

**Comment by rxu** [comment]
Webpush notification won't send for some reason, getting PHP timeout on posting with no errors.

**Comment by iMattPro** [comment]
Maybe it won't push from a localhost? We should probably the unit tests from the master's PR in here at some point?

**Comment by rxu** [comment]
Well, it does push from a localhost with Marc's branch, and actually it tries to push here too but `Brick\Math\Internal\Calculator\BcMathCalculator->divQR` takes too long for some reason.

**Comment by rxu** [comment]
Okay, rechecked and now Marc's code takes long (no timeout though) and doesn't send webpush as well. Probably something's wrong with PHP settings on my side.

**Comment by rxu** [comment]
Enabling PHP `gmp` extension made posting quick and got rid of timeout.

**Comment by rxu** [comment]
And okay, tried to get webpush with this extension from the Edge browser and successful. Not with Opera still.

**Comment by rxu** [comment]
Firefox 121.0 failed to webpush too.

**Comment by rxu** [comment]
As a conclusion, for me it works but unstable.
MS Edge sends webpush every time.
Chrome and Firefox sends webpush after I ran them.
Opera doesn't send webpush at all.

I assume this PR could be merged as in common it did its task ("First partial files set") and the main functionality does work.

**Pull request merged**